From 9c4b11e0a540b8832ac0bf87f88629a13012f2a4 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Thu, 10 Jul 2014 01:06:20 -0400
Subject: [x86] hotplug: Fix online state in cpu0 debug interface

Message-id: <20140710010619.22146.21408.stgit@gir.stowe>
Patchwork-id: 85616
O-Subject: [RHEL7.1 PATCH 078/105] hotplug / x86: Fix online state in cpu0 debug interface
Bugzilla: 1114228
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1114228
Upstream Status: f6913f990273a1fbfc0ec501d4998faa4f1ebae6

commit f6913f990273a1fbfc0ec501d4998faa4f1ebae6
Author: Toshi Kani <toshi.kani@hp.com>
Date:   Thu Aug 29 18:22:06 2013 -0600

    hotplug / x86: Fix online state in cpu0 debug interface

    _debug_hotplug_cpu() is a debug interface that puts cpu0 offline during
    boot-up when CONFIG_DEBUG_HOTPLUG_CPU0 is set.  After cpu0 is put offline
    in this interface, however, /sys/devices/system/cpu/cpu0/online still
    shows 1 (online).

    This patch fixes _debug_hotplug_cpu() to update dev->offline when CPU
    online/offline operation succeeded.

    Signed-off-by: Toshi Kani <toshi.kani@hp.com>
    Acked-by: Yasuaki Ishimatsu <isimatu.yasuaki@jp.fujitsu.com>
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kernel/topology.c b/arch/x86/kernel/topology.c
index 6e60b5f..5823bbd 100644
--- a/arch/x86/kernel/topology.c
+++ b/arch/x86/kernel/topology.c
@@ -72,16 +72,19 @@ int __ref _debug_hotplug_cpu(int cpu, int action)
   ret = cpu_down(cpu);
   if (!ret) {
    pr_info("CPU %u is now offline\n", cpu);
+   dev->offline = true;
    kobject_uevent(&dev->kobj, KOBJ_OFFLINE);
   } else
    pr_debug("Can't offline CPU%d.\n", cpu);
   break;
  case 1:
   ret = cpu_up(cpu);
-  if (!ret)
+  if (!ret) {
+   dev->offline = false;
    kobject_uevent(&dev->kobj, KOBJ_ONLINE);
-  else
+  } else {
    pr_debug("Can't online CPU%d.\n", cpu);
+  }
   break;
  default:
   ret = -EINVAL;
-- 
1.7.1